open-craft/ecommerce
# Add Support for Pluggable Payment Processors

This backports PR https://github.com/edx/ecommerce/pull/3297 to add pluggable payment processors to ecommerce and adds some additional features needed for the HyperPay payment processor backend implemented in [BB-3801](https://tasks.opencraft.com).

**JIRA tickets**: Implements BB-3315

**Dependencies**: None

**Testing instructions**:

A test account needs to be created on paytabs, and the fields filled in the ecommerce repo in `ecommerce/settings/devstack.py` under the `PAYMENT_PROCESSOR_CONFIG`. Also paytabs needs an externally accessible address to finish the payment. We're using ngrok to expose the devstack ecommerce port for paytabs to callback to. Here is an example config stripped of credentials:
```
'paytabs': {
    'merchant_email': 'test@example.com',
    'secret_key': '************************************'
    'default_phone_number_country_code': '00971',
    'return_base_url': 'https://f544d7fd8e8d.ngrok.io'
},

```
* Try setting a `CHECKOUT_TEXT` attribute to a translatable string on the payment processor class and verify that the text shows up in the checkout button for the processor.
* Try adding a custom app (say ecommerce-paytabs) to `ADDL_INSTALLED_APPS` in `ecommerce.yml` and verify that it is added to the `INSTALLED_APPS` Django setting.

**Author notes and concerns**:

1. There isn't a simple way to bubble error messages back to learner's browsers from within the basket page. If all fields are not set in their account profile, they get a generic "transaction failed" when PayTabs refuses if they are missing even one required field, which would make it difficult for them to determine on their own what they need to set in the account profile editor.



**Reviewers**
